### ISSUE TITLE
Callback in "initUserSessionWithCallback" is being called multiple times

### ISSUE THREAD

**Issue by NRajabalee** [BranchMetrics/ios-branch-deep-linking-attribution#15]
I am getting an issue where the callback passed in "initUserSessionWithCallback" is being called each time the app comes to foreground. Why is the callback being called more than once? Should it not be called only once?

When the callback the first time right after clicking on a link, the params list contains  information. For each subsequent call, the params list is usually empty, which is good. However...I've encountered cases where the params list is NOT empty for each subsequent call. This results in my welcome alert view appearing more than once.

This is my set-up:
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
  {
  [Branch getInstance:@"<myAppIDHere>"];
  
  [[Branch getInstance] initUserSessionWithCallback:^(NSDictionary *params) {
         //This callback is being called each time the app comes to foreground.
          NSLog(@"Referral link data: %@", [params description]);
          if (params.count > 0){
             //Show a welcome alert view based on param values. 
          }
  } withLaunchOptions:launchOptions];
  
  return YES;
  }

-(BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation {
    if ([[Branch getInstance] handleDeepLink:url]) {
        return YES;
    }
    return NO;
}

**Comment by erickreutz** [comment]
@aaustin what triggers the init callback being called again? I'm running into problems when it's called multiple times at what seems to be random points during the applications life when it has not been closed or re-opened.

**Comment by shortstuffsushi** [comment]
Hi @erickreutz, there are currently four places we initialize the session.

* [When you call initSession](https://github.com/BranchMetrics/Branch-iOS-SDK/blob/bd888ce6678792efdda2ca90750465cdf09b4d82/Branch-SDK/Branch-SDK/Branch.m#L333)
* [When a Deep Link is handled](https://github.com/BranchMetrics/Branch-iOS-SDK/blob/bd888ce6678792efdda2ca90750465cdf09b4d82/Branch-SDK/Branch-SDK/Branch.m#L352)
* [When you attempt to make an API request before calling init](https://github.com/BranchMetrics/Branch-iOS-SDK/blob/bd888ce6678792efdda2ca90750465cdf09b4d82/Branch-SDK/Branch-SDK/Branch.m#L380)
* [When the application becomes active](https://github.com/BranchMetrics/Branch-iOS-SDK/blob/bd888ce6678792efdda2ca90750465cdf09b4d82/Branch-SDK/Branch-SDK/Branch.m#L963)

All of these will only occur if an Init request isn't already in progress / enqueued. Note that if you make an API request before calling init, we don't attempt to call the init callback.

If you're seeing it mid-run, that wouldn't be expected behavior unless init had failed earlier, in which case it shouldn't be calling your callback again. Could you provide us with the version you're running as well?

**Comment by TejasviTandon** [comment]
I am facing the similar problem. I initialise the Branch session in appDidFinishLaunching() and after that I navigate user to Login screen. On Login screen user authenticates himself using Biometric. Everything fine till here.
Now as soon as the Local authentication is completed, initUserSessionWithCallback is getting called again.
Note: My branch initialisation is getting Failed too.

**Comment by shortstuffsushi** [comment]
Hey @TejasviTandon - as this issue is nearing five years old, and a lot has changed in the SDK since that time, I would recommend opening a new issue with some additional detail. It sounds like you may be having a different problem, since your init call is failing, which is likely going to intentionally cause a second init to take place.
